proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,733 @@
+module declarations within modules. The syntax for a local module is the same as for a file
+module, only indented appropriately for the current scope.
+
+Local modules:
+
+- Can have ``import module`` statements, but not normal ``import`` statements.
+- Can have nested module declarations.
+- Can have export specifiers.
+
+Local module declarations mostly behave the same as file modules:
+
+- A local module declaration adds a binding for its declared name to
+  its enclosing scope, as usual for declarations.
+- Bindings from the enclosing scope are in scope in the local module,
+  but are *shadowed* if there is a local declaration of the same name. [#]_
+
+.. [#] See `Recursion`_ for some subtleties.
+
+Additionally, local modules cannot define typeclass instances or type family
+instances. [#]_
+
+.. [#] See `No local typeclass or type family instances - reasoning`_ for reasoning.
+
+Module aliases
+^^^^^^^^^^^^^^
+
+We add the language extension ``ModuleAliases``. This allows a
+new kind of top-level declaration: ``module <name> = <qualname>``.
+
+A module alias adds a binding in the current scope from the new name
+to the module referred to by the right hand side.
+
+Examples
+--------
+
+Exporting small utility modules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+::
+
+   module A (f, module Unsafe) where
+
+   f = Unsafe.g …
+
+   module Unsafe where
+       g = … 
+
+Exporting qualified names
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal allows exporting modules, but doesn’t say they can be
+exported under *qualified* names, which is something that people want to
+do. Consider the following `proposed
+example <https://github.com/ghc-proposals/ghc-proposals/pull/283#issuecomment-541525245>`_:
+
+::
+
+   module MyPrelude ( qualified module BL
+                    , qualified module BS
+                    , Set
+                    , qualified module Set ) where
+
+
+   import qualified Data.ByteString.Lazy as BL
+   import qualified Data.ByteString as BS
+   import Data.Set ( Set )
+   import qualified Data.Set as Set
+
+The goal of this is to re-export everything in `Data.ByteString.Lazy` qualified
+under `BL`, so for example `BL.ByteString` would be `Bytestring` from `Data.Bytestring.Lazy`.
+And similarly for the other modules.
+
+This could be achieved as follows:
+
+::
+
+   module MyPrelude (module BL, module BS, Set, module Set) where
+
+   import qualified Data.ByteString.Lazy as BL
+   import qualified Data.ByteString as BS
+   import Data.Set ( Set )
+   import qualified Data.Set as Set
+
+This example is simple, because we only want a single level of qualification.
+For example, suppose that we instead want to re-export `Data.Bytestring.Lazy`
+as `Data.Bytestring`. Then we need to write:
+
+::
+
+   module MyPrelude (module Data) where
+
+   import qualified Data.ByteString.Lazy as BL
+
+   module Data (module ByteString) where
+       module ByteString = BL
```

Comment:
How does this differ from just importing "DBL" as "DB", already possible today?
```
$ ghci
GHCi, version 8.6.3: http://www.haskell.org/ghc/  :? for help
Prelude> import Data.ByteString.Lazy as Data.ByteString
Prelude Data.ByteString> :t Data.ByteString.toStrict 
Data.ByteString.toStrict
  :: ByteString -> Data.ByteString.Internal.ByteString
```
Editing the tree at intermediate nodes looks like a rather tricky business to me, the dots in the middle of module names don't imply any sort of parent/child or scoping relationships today, and introducing that could make things rather more fragile...

Reply:
> the dots in the middle of module names don't imply any sort of parent/child or scoping relationships today, and introducing that could make things rather more fragile...

Yes, well, that's precisely the thing that this proposal wants to change. In what way would this make things fragile?